sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraServiceImpl.java

```diff
@@ -0,0 +1,317 @@
+    this.port = port;
+    this.keyspace = keyspace;
+    this.table = table;
+  }
+
+  private class CassandraReaderImpl<T> implements Reader<T> {
+
+    private final String splitQuery;
+    private final Class<T> entity;
+    private Cluster cluster;
+    private Session session;
+    private ResultSet resultSet;
+    private Iterator<T> iterator;
+    private T current;
+
+    public CassandraReaderImpl(String splitQuery, Class<T> entity) {
+      this.splitQuery = splitQuery;
+      this.entity = entity;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOG.debug("Starting Cassandra reader");
+      cluster = getCluster();
+      session = cluster.connect();
+      LOG.debug("Query: " + splitQuery);
+      resultSet = session.execute(splitQuery);
+
+      final MappingManager mappingManager = new MappingManager(session);
+      Mapper mapper = mappingManager.mapper(entity);
+      iterator = mapper.map(resultSet).iterator();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() throws IOException {
+      if (iterator.hasNext()) {
+        current = iterator.next();
+        return true;
+      }
+      current = null;
+      return false;
+    }
+
+    @Override
+    public void close() {
+      LOG.debug("Closing Cassandra reader");
+      if (session != null) {
+        session.close();
+      }
+      if (cluster != null) {
+        cluster.close();
+      }
+    }
+
+    @Override
+    public T getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+  }
+
+  @Override
+  public Reader<T> createReader(String query, Class<T> entity) throws IOException {
+    return new CassandraReaderImpl<>(query, entity);
+  }
+
+  @Override
+  public long getEstimatedSizeBytes() {
+    try (Cluster cluster = getCluster()) {
+      List<TokenRange> tokenRanges = getTokenRanges(cluster, keyspace, table);
+      long size = 0L;
+      for (TokenRange tokenRange : tokenRanges) {
+        size += tokenRange.meanPartitionSize * tokenRange.partitionCount;
+      }
+      return Math.round(size / getRingFraction(cluster, tokenRanges));
+    }
+  }
+
+  @Override
+  public List<BoundedSource<T>> splitIntoBundles(
+      CassandraIO.Read<T> spec,
+      long desiredBundleSizeBytes) {
+    long numSplits = 1;
+    List<BoundedSource<T>> sourceList = new ArrayList<>();
+    try {
+      if (desiredBundleSizeBytes > 0) {
+        numSplits = getEstimatedSizeBytes() / desiredBundleSizeBytes;
+      }
+    } catch (Exception e) {
+      LOG.warn("EstimatedSizeBytes / DesiredBundleSizeBytes failed, fallback to an unique split",
+          e);
+      numSplits = 1;
+    }
+    if (numSplits <= 0) {
+      LOG.warn("Number of splits is less than 0 ({}), fallback to 1", numSplits);
+      numSplits = 1;
+    }
+
+    LOG.debug("Number of splits is {}", numSplits);
+
+    if (numSplits == 1) {
```

Comment:
I still think this special case is unnecesasary. This is extra code, and an extra codepath to debug if something goes wrong, and I doubt that it gives a serious performance benefit by not using the Partitioner.